File: frontend/src/app/services/config.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';

export interface FirebaseClientConfig {
  apiKey: string;
  authDomain: string;
  projectId: string;
  storageBucket: string;
  messagingSenderId: string;
  appId: string;
  measurementId?: string;
}

@Injectable({
  providedIn: 'root'
})
export class ConfigService {
  private firebaseConfigCache: FirebaseClientConfig | null = null;
  private configPromise: Promise<FirebaseClientConfig> | null = null;
  private debug(...args: unknown[]): void {
    if (!environment.debugMode || !args.length) {
      return;
    }
    console.log('[ConfigService]', ...args);
  }

  constructor() {}

  async getFirebaseConfig(): Promise<FirebaseClientConfig> {
    // Return cached config if available
    if (this.firebaseConfigCache) {
      return this.firebaseConfigCache;
    }

    // Return existing promise if one is in progress
    if (this.configPromise) {
      return this.configPromise;
    }

    // Create new promise to fetch config
    this.configPromise = this.fetchFirebaseConfig();
    return this.configPromise;
  }

  private async fetchFirebaseConfig(): Promise<FirebaseClientConfig> {
    try {
      this.debug('Fetching Firebase configuration from backend...');

      const response = await fetch('/api/config/firebase', {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
        cache: 'default'
      });

      if (!response.ok) {
        throw new Error(`Failed to fetch Firebase config: ${response.status} ${response.statusText}`);
      }

      const config: FirebaseClientConfig = await response.json();

      // Validate the configuration
      if (!this.isFirebaseConfigValid(config)) {
        throw new Error('Invalid Firebase configuration received from backend');
      }

      // Cache the configuration
      this.firebaseConfigCache = config;
      this.debug('Firebase configuration loaded successfully');

      return config;
    } catch (error) {
      console.error('Failed to fetch Firebase configuration:', error);

      this.configPromise = null;

      console.error('Backend configuration required - no fallback available');
      throw error;
    }
  }

  private isFirebaseConfigValid(config: any): config is FirebaseClientConfig {
    return config &&
           typeof config.apiKey === 'string' &&
           typeof config.authDomain === 'string' &&
           typeof config.projectId === 'string' &&
           typeof config.storageBucket === 'string' &&
           typeof config.messagingSenderId === 'string' &&
           typeof config.appId === 'string' &&
           config.apiKey.length > 0 &&
           config.authDomain.length > 0 &&
           config.projectId.length > 0 &&
           config.storageBucket.length > 0 &&
           config.messagingSenderId.length > 0 &&
           config.appId.length > 0;
  }

  async waitForFirebaseConfig(): Promise<FirebaseClientConfig> {
    return this.getFirebaseConfig();
  }

  isFirebaseConfigValidSync(config: FirebaseClientConfig | null): boolean {
    return config !== null && this.isFirebaseConfigValid(config);
  }

  clearCache(): void {
    this.firebaseConfigCache = null;
    this.configPromise = null;
  }
}

File: backend/cmd/server/config_service.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
	"strings"
)

// FirebaseClientConfig contains the client-side Firebase configuration
type FirebaseClientConfig struct {
	APIKey            string `json:"apiKey"`
	AuthDomain        string `json:"authDomain"`
	ProjectID         string `json:"projectId"`
	StorageBucket     string `json:"storageBucket"`
	MessagingSenderID string `json:"messagingSenderId"`
	AppID             string `json:"appId"`
	MeasurementID     string `json:"measurementId"`
}

// ConfigService manages application configuration
type ConfigService struct {
	firebaseConfig       FirebaseClientConfig
	corsConfig           CORSConfig
	stripePublishableKey string
}

// CORSConfig holds CORS configuration
type CORSConfig struct {
	AllowedOrigins []string
	AllowedMethods []string
	AllowedHeaders []string
	MaxAge         int
}

// NewConfigService creates a new configuration service
func NewConfigService() *ConfigService {
	return &ConfigService{
		firebaseConfig:       loadFirebaseConfig(),
		corsConfig:           loadCORSConfig(),
		stripePublishableKey: loadStripePublishableKey(),
	}
}

// loadStripePublishableKey loads Stripe publishable key from environment
func loadStripePublishableKey() string {
	key := os.Getenv("STRIPE_PUBLISHABLE_KEY")
	if key == "" {
		log.Println("⚠️  STRIPE_PUBLISHABLE_KEY not set - Stripe functionality will be unavailable")
	}
	return key
}

// loadFirebaseConfig loads Firebase configuration from environment variables
func loadFirebaseConfig() FirebaseClientConfig {
	// Load all Firebase configuration from environment variables
	config := FirebaseClientConfig{
		APIKey:            os.Getenv("FIREBASE_API_KEY"),
		AuthDomain:        os.Getenv("FIREBASE_AUTH_DOMAIN"),
		ProjectID:         os.Getenv("FIREBASE_PROJECT_ID"),
		StorageBucket:     os.Getenv("FIREBASE_STORAGE_BUCKET"),
		MessagingSenderID: os.Getenv("FIREBASE_MESSAGING_SENDER_ID"),
		AppID:             os.Getenv("FIREBASE_APP_ID"),
		MeasurementID:     os.Getenv("FIREBASE_MEASUREMENT_ID"),
	}

	// Validate required fields (frontend needs these for Firebase SDK initialization)
	var missing []string
	if config.APIKey == "" {
		missing = append(missing, "FIREBASE_API_KEY")
	}
	if config.AuthDomain == "" {
		missing = append(missing, "FIREBASE_AUTH_DOMAIN")
	}
	if config.ProjectID == "" {
		missing = append(missing, "FIREBASE_PROJECT_ID")
	}
	if config.StorageBucket == "" {
		missing = append(missing, "FIREBASE_STORAGE_BUCKET")
	}
	if config.MessagingSenderID == "" {
		missing = append(missing, "FIREBASE_MESSAGING_SENDER_ID")
	}
	if config.AppID == "" {
		missing = append(missing, "FIREBASE_APP_ID")
	}

	if len(missing) > 0 {
		log.Fatalf("❌ Firebase configuration error: Missing required environment variables: %v\n"+
			"Frontend requires these for Firebase SDK initialization.\n"+
			"Please set all Firebase variables in your .env file", missing)
	}

	log.Printf("✅ Firebase configuration loaded successfully (Project: %s)", config.ProjectID)
	return config
}

// loadCORSConfig loads CORS configuration from environment variables
func loadCORSConfig() CORSConfig {
	allowedOrigins := envOr("ALLOWED_ORIGINS", "http://localhost:4200,https://localhost:4200")

	origins := strings.Split(allowedOrigins, ",")
	if os.Getenv("ENVIRONMENT") == "production" {
		origins = filterLocalhostOrigins(origins)
	}

	return CORSConfig{
		AllowedOrigins: origins,
		AllowedMethods: []string{"GET", "POST", "PUT", "DELETE", "OPTIONS"},
		AllowedHeaders: []string{"Authorization", "Content-Type", "X-Requested-With"},
		MaxAge:         3600,
	}
}

// filterLocalhostOrigins removes localhost origins for production security
func filterLocalhostOrigins(origins []string) []string {
	var filtered []string
	for _, origin := range origins {
		origin = strings.TrimSpace(origin)
		if !strings.Contains(origin, "localhost") && !strings.Contains(origin, "127.0.0.1") {
			filtered = append(filtered, origin)
		} else {
			log.Printf("Filtered out localhost origin in production: %s", origin)
		}
	}

	if len(filtered) == 0 {
		log.Fatal("No valid origins configured for production. Please set ALLOWED_ORIGINS environment variable.")
	}

	return filtered
}

// GetFirebaseConfig returns the safe Firebase configuration for clients
func (cs *ConfigService) GetFirebaseConfig() FirebaseClientConfig {
	return cs.firebaseConfig
}

// GetCORSConfig returns the CORS configuration
func (cs *ConfigService) GetCORSConfig() CORSConfig {
	return cs.corsConfig
}

// IsFirebaseConfigValid validates that the Firebase configuration is complete
func (cs *ConfigService) IsFirebaseConfigValid() bool {
	config := cs.firebaseConfig
	return config.ProjectID != "" &&
		   config.AuthDomain != "" &&
		   config.StorageBucket != "" &&
		   config.MessagingSenderID != "" &&
		   config.AppID != ""
}

// handleFirebaseConfig serves the safe Firebase configuration to clients
func (cs *ConfigService) handleFirebaseConfig(w http.ResponseWriter, r *http.Request) {
	if r.Method != "GET" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "public, max-age=3600")
	w.Header().Set("X-Content-Type-Options", "nosniff")

	if err := json.NewEncoder(w).Encode(cs.firebaseConfig); err != nil {
		log.Printf("Error encoding Firebase config: %v", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	log.Printf("Served Firebase configuration to client")
}

// handleStripeConfig serves the Stripe publishable key to clients
func (cs *ConfigService) handleStripeConfig(w http.ResponseWriter, r *http.Request) {
	if r.Method != "GET" {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Cache-Control", "public, max-age=3600")
	w.Header().Set("X-Content-Type-Options", "nosniff")

	response := map[string]string{
		"publishableKey": cs.stripePublishableKey,
	}

	if err := json.NewEncoder(w).Encode(response); err != nil {
		log.Printf("Error encoding Stripe config: %v", err)
		http.Error(w, "Internal server error", http.StatusInternalServerError)
		return
	}

	log.Printf("Served Stripe configuration to client")
}

